python/pyarrow/scalar.pxi

```diff
@@ -73,6 +73,12 @@ cdef class ArrayValue(Scalar):
             raise NotImplementedError(
                 "Cannot compare Arrow values that don't support as_py()")
 
+    def __hash__(self):
```

Comment:
Ok, I think we misunderstood each other.
When you asked about restricting yourself to integers, I thought you meant about writing a fast path that avoids calling `as_py()`. If there is no fast path, then there's no need to check for ints or any other types.

Reply:
Sorry about that. I've switch it to `hash()` using `as_py()` without any type checks. Per the JIRA ticket can we consider a separate ticket for a fast path custom hash on ints?

Reply:
> Per the JIRA ticket can we consider a separate ticket for a fast path custom hash on ints?

Definitely.

Reply:
Ticket created https://issues.apache.org/jira/browse/ARROW-2339. Anything else you would suggest for this PR?